File: lab1_ciphers/c_vernam.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <arpa/inet.h>

#define PORT 8080
#define BUFFER_SIZE 1024

// Vernam cipher decryption function
void decrypt(char *ciphertext, char *key, char *plaintext)
{
    int len = strlen(ciphertext);
    for (int i = 0; i < len; i++)
    {
        plaintext[i] = ((ciphertext[i] ^ key[i]) % 26) + 'A';
    }
    plaintext[len] = '\0';
}

int main()
{
    int sock = 0;
    struct sockaddr_in serv_addr;
    char ciphertext[BUFFER_SIZE] = {0};
    char key[BUFFER_SIZE] = {0};
    char plaintext[BUFFER_SIZE];

    // Creating socket
    if ((sock = socket(AF_INET, SOCK_STREAM, 0)) < 0)
    {
        perror("Socket creation error");
        return -1;
    }

    serv_addr.sin_family = AF_INET;
    serv_addr.sin_port = htons(PORT);

    // Converting IP address to binary
    if (inet_pton(AF_INET, "127.0.0.1", &serv_addr.sin_addr) <= 0)
    {
        perror("Invalid address or address not supported");
        return -1;
    }

    // Connecting to server
    if (connect(sock, (struct sockaddr *)&serv_addr, sizeof(serv_addr)) < 0)
    {
        perror("Connection failed");
        return -1;
    }

    // Receiving ciphertext and key
    int valread = read(sock, ciphertext, BUFFER_SIZE);
    ciphertext[valread] = '\0';

    valread = read(sock, key, BUFFER_SIZE);
    key[valread] = '\0';

    printf("Ciphertext received: %s\n", ciphertext);
    printf("Key received: %s\n", key);

    // Decrypt the ciphertext
    decrypt(ciphertext, key, plaintext);
    printf("Decrypted message: %s\n", plaintext);

    close(sock);
    return 0;
}

File: lab1_ciphers/c_vernam.py
import socket

PORT = 8080
BUFFER_SIZE = 1024

# Vernam cipher decryption function
def decrypt(ciphertext, key):
    plaintext = ''.join(chr(((ord(c) ^ ord(k)) % 26) + ord('A')) for c, k in zip(ciphertext, key))
    return plaintext

client_socket = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
client_socket.connect(("127.0.0.1", PORT))

# Receive full message
data = client_socket.recv(BUFFER_SIZE).decode()

# Ensure the message is correctly split into ciphertext and key
if "::" in data:
    ciphertext, key = data.split("::")
else:
    print("Error: Message format incorrect.")
    client_socket.close()
    exit(1)

print(f"Ciphertext received: {ciphertext}")
print(f"Key received: {key}")

# Decrypt the ciphertext
plaintext = decrypt(ciphertext, key)
print(f"Decrypted message: {plaintext}")

client_socket.close()
